R/shiny.R

```diff
@@ -595,5 +597,5 @@ read_shiny_deps <- function(files_dir) {
 # shiny:::setCurrentTheme() was added in 1.6 (we may export in next version)
 set_current_theme <- function(theme) {
   set_theme <- asNamespace("shiny")$setCurrentTheme
-  if (is.function(set_theme)) set_theme(theme)
+  if (is.function(set_theme) && is_bs_theme(theme)) set_theme(theme)
```

Comment:
Sorry for the bad advice/confusion @cderv -- What I had before was actually correct. Even when we're not using bslib, we still want to "clear" the previously set theme (if any) so that future theming code knows that bslib is not actually relevant. I was essentially worried about [themable components](https://rstudio.github.io/bslib/articles/theming.html#dynamically-themeable-component) using an `is.null(theme)`/`length(theme)` to check for the presence of a bslib theme (when they should be using `is_bs_theme(theme)`, but after thinking about it more, I think this is the right way to go:

```suggestion
  if (is.function(set_theme)) set_theme(theme)
```

Reply:
I see - I'll revert then. 
However, my question still remains from this comment discussion: https://github.com/rstudio/rmarkdown/pull/2065#discussion_r586560187

Basically, we are retrieving the `theme` information from the YAML. If there is none `theme` passed to `set_current_theme ` is NULL **but** if the output format as a default theme base on **bslib**, then you won't have that information. Also, `html_document` default theme is `"default"`, but this is information is not in the YAML but in the output format itself. 

```
format <- output_format_from_yaml_front_matter(
    input_lines = read_utf8(target_file),
    output_options = render_args$output_options
)
# will be null is no `theme` in the YAML or in `render_args$output_options`
format$options$theme

I am under the impression you need to know the theme set and used in the document, to make shiny aware. 
If so, I think we are missing something. 

Does that make sense ?

Reply:
Let me rephrase to see if I understand correctly: you're worried about the case where no `theme` is specified in the `target_file`'s yaml or the `render_args`  (in that case, `format$options$theme` will be `NULL` even though it is really `formals(output_format)$theme`, right)? In theory, yes, this is a problem (in practice, it's not currently a big problem since no formats default to a `{bslib}` theme).  That being said, if it'd be great if we did do the right thing here (something like `theme <- format$options$theme %||% formals(output_format)$theme` -- do you know how the get the `output_format` function?)

Reply:
I think you rephrased correctly and this was my thoughts:

> In theory, yes, this is a problem (in practice, it's not currently a big problem since no formats default to a {bslib} theme).

I think we could have in the near future some output format that would default do **bslib** theme maybe. 
But on second thought, if this happens it would be more complicated. Usually, the theme argument is processed inside the output formats - e.g. a now format that would default to pulse bootswatch theme.
```r
if (theme == "default") theme = bslib::bs_theme(bootswatch = "pulse")
```
In this case, it is not easy to retrieve the information, and `formals(output_format)$theme` would be `"default"` which may not help. 

So in practice as you said, it is not a big problem currently. We may leave as is and think more of this later when / if new formats with bslib theme as default come up ?

Reply:
Ahh, right, good point. Let's add a comment about this scenario just above the `set_current_theme()` call. I think perhaps the best way to solve this issue would be to `set_current_theme()` again inside of `html_document_base()` before any user code runs. Is it possible to detect a shiny runtime the `pre_knit` hook?

Reply:
> Is it possible to detect a shiny runtime the pre_knit hook?

Currently it is not supported - runtime is known but not passed in pre_knit processor function, only in post_knit processor function. 

But this is something we can easily add IMO - if developers have followed advice in the doc: 
> pre_knit: An optional function that runs before knitting which receives the input (input filename passed to render) and **... (for future expansion)** arguments.

If unfortunately so custom formats use a pre_knit function without ..., then it will be difficult to add.

Reply:
I'd be all for it if you're happy to deal with pain it might cause :)

Reply:
I'll look into it. 

To summarize the issue we are trying to solve, theme info must be passed to shiny before knit so that customization can by user in a chunk (for bslib themer specically). Otherwise, if shiny does not have the information of the document theme it won't work correctly. 

Did I understand correctly ?

Reply:
yep